File: kiosk-flutter/lib/models/ticket.dart
class Ticket {
  final int id;
  final String numero;
  final int centreId;
  final int? serviceId;
  final String type; // 'rdv' ou 'sans_rdv'
  final String statut; // 'en_attente', 'appelé', 'terminé', 'absent'
  final String? heureRdv;
  final int priorite;
  final DateTime createdAt;
  final DateTime? calledAt;
  final DateTime? completedAt;
  final int? guichetId;

  Ticket({
    required this.id,
    required this.numero,
    required this.centreId,
    this.serviceId,
    required this.type,
    required this.statut,
    this.heureRdv,
    required this.priorite,
    required this.createdAt,
    this.calledAt,
    this.completedAt,
    this.guichetId,
  });

  factory Ticket.fromJson(Map<String, dynamic> json) {
    return Ticket(
      id: json['id'] as int,
      numero: json['numero'] as String,
      centreId: json['centre_id'] as int,
      serviceId: json['service_id'] as int?,
      type: json['type'] as String,
      statut: json['statut'] as String,
      heureRdv: json['heure_rdv'] as String?,
      priorite: json['priorite'] as int? ?? 1,
      createdAt: DateTime.parse(json['created_at'] as String),
      calledAt: json['called_at'] != null 
          ? DateTime.parse(json['called_at'] as String)
          : null,
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'] as String)
          : null,
      guichetId: json['guichet_id'] as int?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'numero': numero,
      'centre_id': centreId,
      'service_id': serviceId,
      'type': type,
      'statut': statut,
      'heure_rdv': heureRdv,
      'priorite': priorite,
      'created_at': createdAt.toIso8601String(),
      'called_at': calledAt?.toIso8601String(),
      'completed_at': completedAt?.toIso8601String(),
      'guichet_id': guichetId,
    };
  }
}

File: app/Models/Ticket.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Ticket extends Model
{
    protected $fillable = [
        'numero',
        'centre_id',
        'service_id',
        'user_id',
        'guichet_id',
        'statut',
        'type',
        'priorite',
        'heure_rdv',
        'called_at',
        'completed_at',
        'metadata',
        'client_id',
        'retrait_type_piece',
        'retrait_numero_recepisse',
        'retrait_scan_recepisse',
        'retrait_numero_piece_finale',
        'retrait_date_expiration_piece',
    ];

    protected $casts = [
        'heure_rdv' => 'datetime:H:i',
        'called_at' => 'datetime',
        'completed_at' => 'datetime',
        'metadata' => 'array',
        'retrait_date_expiration_piece' => 'date',
    ];

    // Constantes pour les statuts
    const STATUT_EN_ATTENTE = 'en_attente';
    const STATUT_APPELÉ = 'appelé';
    const STATUT_EN_COURS = 'en_cours';
    const STATUT_TERMINÉ = 'terminé';
    const STATUT_ABSENT = 'absent';
    const STATUT_ANNULÉ = 'annulé';
    const STATUT_EN_ATTENTE_BIOMETRIE = 'en_attente_biometrie';
    const STATUT_EN_COURS_BIOMETRIE = 'en_cours_biometrie';

    // Constantes pour les types
    const TYPE_RDV = 'rdv';
    const TYPE_SANS_RDV = 'sans_rdv';

    /**
     * Scope pour les tickets en attente
     */
    public function scopeEnAttente($query)
    {
        return $query->where('statut', self::STATUT_EN_ATTENTE);
    }

    /**
     * Scope pour les tickets appelés
     */
    public function scopeAppelé($query)
    {
        return $query->where('statut', self::STATUT_APPELÉ);
    }

    /**
     * Scope pour les tickets du jour
     */
    public function scopeDuJour($query)
    {
        return $query->whereDate('created_at', today());
    }

    /**
     * Scope pour un centre donné
     */
    public function scopePourCentre($query, $centreId)
    {
        return $query->where('centre_id', $centreId);
    }

    /**
     * Scope pour ordonner par priorité puis date
     */
    public function scopeParPriorite($query)
    {
        return $query->orderBy('priorite', 'desc')->orderBy('created_at', 'asc');
    }

    public function centre()
    {
        return $this->belongsTo(Centre::class);
    }

    public function service()
    {
        return $this->belongsTo(Service::class);
    }

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function guichet()
    {
        return $this->belongsTo(Guichet::class);
    }

    public function client()
    {
        return $this->belongsTo(Client::class);
    }

    public function retraitCarte()
    {
        return $this->hasOne(RetraitCarte::class);
    }
}
